### PR TITLE
Green tea m us3

### PR DESCRIPTION
**Pull request by PiotrKowandy**

### CODE REVIEW

File: src/CourseDashboard.hpp [hunk 1]
```diff
@@ -12,8 +12,9 @@ class CourseDashboard
     void deleteUserByNick(std::string nick);
     void retriveUserByNick(std::string nick);
     void updateUser(User & user);
+    bool logIn();
 
 private:
     std::vector<User> users_;
-
+    std::map<std::string, std::string> map_;
```
Comment (ziobron) [inline src/CourseDashboard.hpp]: the name map_ is not meaningful

File: src/User.cpp [hunk 2]
```diff
@@ -22,6 +26,8 @@ std::string User::getAllInfo() const
     ss<<std::setw(1) << "3." << group_ << "|";
     ss<<std::setw(1) << "4." << gitHub_ << "|";
     ss<<std::setw(1) << "5." << firecode_ << "|";
+    ss<<std::setw(1) << "6." << email_ << "|";
+    ss<<std::setw(1) << "7." << password_ << "|";
```
Comment (ziobron) [inline src/User.cpp]: printing password is not a good idea

File: tests/CourseDashboardTests.cpp [hunk 3]
```diff
@@ -1,8 +1,35 @@
 #include <gtest/gtest.h>
 #include "CourseDashboard.hpp"
+#include "User.hpp"
 
+class CourseDashboardTests : public ::testing::Test
+{
+protected:
+    CourseDashboard* testCdb;
+    User* testU1;
```
Comment (ziobron) [inline tests/CourseDashboardTests.cpp]: why pointers, not normal values?

File: tests/CourseDashboardTests.cpp [hunk 4]
```diff
@@ -1,8 +1,35 @@
...
+    {
+        testU1 = new User("Szymon", "SzymonGajewski", "wieczorowa", "SzymonGajewski", "SzymonGajewski", "mail@com", "password");
+        testCdb = new CourseDashboard();
+    }
+
+    CourseDashboardTests()
+    {
+        EXPECT_NE("mail", testU1->getEmail());
+        EXPECT_EQ("mail@com", testU1->getEmail());
+        EXPECT_TRUE(testCdb->logIn());
```
Comment (ziobron) [inline tests/CourseDashboardTests.cpp]: how is user input tested? logIn() uses cin.

File: tests/CourseDashboardTests.cpp [hunk 5]
```diff
@@ -1,8 +1,35 @@
...
+        delete testU1;
+        delete testCdb;
+    }
+};
 
 TEST(CourseDashboardTests, canConstructCourseDashboard)
 {
     CourseDashboard cd;
 }
+
```
Comment (ziobron) [inline tests/CourseDashboardTests.cpp:33]: what about the test for failed login?

**Review by ziobron — CHANGES_REQUESTED** [review]

File: src/CourseDashboard.hpp (hunk 1 above)
Comment (PiotrKowandy) [reply src/CourseDashboard.hpp]: corrected:
    std::map<std::string, std::string> usersLoginMap_;

File: tests/CourseDashboardTests.cpp (hunk 3 above)
Comment (PiotrKowandy) [reply tests/CourseDashboardTests.cpp]: I thought this variable will be more powerful:), but without deeper sense. Cause we don't use polymorphism yet.

File: src/User.cpp (hunk 2 above)
Comment (PiotrKowandy) [reply src/User.cpp]: Corrected, Line with password has been removed.

File: tests/CourseDashboardTests.cpp (hunk 3 above)
Comment (PiotrKowandy) [reply tests/CourseDashboardTests.cpp]: Currently there is no any pointers in tests.

File: tests/CourseDashboardTests.cpp (hunk 5 above)
Comment (PiotrKowandy) [reply tests/CourseDashboardTests.cpp:33]: Two separate log in tests deployed, one related to success, second one related to fail.

File: tests/CourseDashboardTests.cpp (hunk 4 above)
Comment (PiotrKowandy) [reply tests/CourseDashboardTests.cpp]: All EXPECTs swapped to ASSERT_TRUE.

Comment (PiotrKowandy) [reply tests/CourseDashboardTests.cpp]: New class has been created to derive logIn function from CourseDashboard. In test function we are able to set password and login. After that we can use redefined logIn() function without any cin.

File: src/CourseDashboard.cpp [hunk 6]
```diff
@@ -82,3 +84,40 @@ void CourseDashboard::updateUser(User & user)
...
+void CourseDashboard::addTrainer(std::string login)
+{
+    trainersLoginVector_.emplace_back(login);
+}
+
+bool CourseDashboard::logIn()
+{
+    std::string loginKeyword;
+    std::cout << "Enter  e-mail/name : ";
+    std::cin >> loginKeyword;
```
Comment (ziobron) [inline src/CourseDashboard.cpp:97]: The trainer can log in without a password?

File: src/CourseDashboardLogIn.hpp [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+public:
+    CourseDashboardLogIn();
+    ~CourseDashboardLogIn();
+    bool setLoginPassword(std::string, std::string);
+    std::string getLog();
+    std::string getPass();
+    bool logIn();
+
+private:
+    std::string log_, pass_;
```
Comment (ziobron) [inline src/CourseDashboardLogIn.hpp:16]: why not login_ and password_?

File: src/CourseDashboardLogIn.cpp [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+#include <stdexcept>
+#include <iostream>
+#include <algorithm>
+#include <fstream>
+
+CourseDashboardLogIn::CourseDashboardLogIn()
+{}
+
+CourseDashboardLogIn::~CourseDashboardLogIn()
+{}
```
Comment (ziobron) [inline src/CourseDashboardLogIn.cpp:11]: Empty constructor and destructor will be generated anyway, they should not be implemented.

File: src/CourseDashboardLogIn.cpp [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+        if (searchUser != usersLoginMap_.end()) 
+        {
+            std::string passwordKeyword = this->getPass();
+            return ((searchUser->second) == passwordKeyword);
+        }
+        else
+        {
+            return false;
+        }
+    }
```
Comment (ziobron) [inline src/CourseDashboardLogIn.cpp:53]: This function looks very similar to login() from CourseDashboard class. Code duplication?

**Comment by ziobron** [comment]
+8 points. Trainer has some other logging logic which was not tested.

**Pull request closed without merging**